Performance improvements and corrected etag for lang.js connector #modxbughunt

wuuti opened the pull request:
### What does it do ?

This PR attempts to improve the performance and caching logic for the lang.js files created by connector. It also fixes an issue with etags causing never a 304 modified being sent. 

It is quite unlogical to regenerate the dictionary (=lexicon entries collection) for every manager page request according to _some_ topics maybe needed. As a result, nearly every page generates its own dictionary. Why don't we send the _whole_ dictionary for a language and the system at once to the browser? It is send only once, after that the browser does not need to reload it regardless on which page you are. 
In this PR the dictionary is generated by getting all namespaces and all topics inside, then fetching the whole lexicon for all topics. The result is cached (currently in cache partition 'lexicon-js'), so the etag computation can be done efficiently. 
### Why is it needed ?

Speeds up manager experience and fixes bug with etag-header. The requests to connectors/lang.js.php get a lot faster and re-requests get the correct 304 not modified.
### Open issues/enhancements
- purge cache partition on cache refresh and lexicon changes for that language. 
- (update) the cache partition also needs to be refreshed when an extra is installed/updated

Commit:
Performance improvements with caching and corrected etag.

Commit:
Performance improvements with caching and corrected etag.

Commit:
Merge remote-tracking branch 'upstream/2.4.x' into improve-lexicon

Commit:
Bug fix with etag.
Added cache partition refresh.

Commit:
Bug fix with etag.
Added cache partition refresh.

inreti commented:
Did some updates to the PR. When cache refresh runs, the cache for the dictionary is refreshed too.
Note: this does not necessarily mean that on the next load a 200 is sent - if the newly generated dictionary hash matches again the browsers etag, a 304 is sent anyway, because the dictionary contents have not changed. So cache refresh only triggers a new dictionary load by the browser if something has really changed in the lexicon topics.

Mark-H commented:
This is interesting. I'm not 100% sold yet on the idea of including all namespaces and all topics - that could be a shit ton (technical term) of data that never gets used. Even if its cached, is it beneficial on say a mobile device (2.5 is more mobile friendly after all), or is this going to increase the download size on that first hit by so much that their out of their monthly quota? 

Any feedback from @modxcms/integration-managers-revolution?

Mark-H commented:
I've tried this patch earlier on the modmore site (which has, for reasons including legacy and a lot of custom stuff, _48_ namespaces) and the lang.js.php file went from 50kb (15kb gzipped) to 498kb (123kb gzipped) so it's approximately 10 times bigger in this particular example. 

In the bigger picture of the manager that's actually not too terrible though, and the caching is a nice bonus. Just wondering if there's any issues with loading all namespaces and topics..

inreti commented:
@Mark-H I am quite sure the data can be handled even by small mobile devices. If you dare to use the manager on those devices, these additional Bytes won't harm you compared to all the other data transferred. And most installations won't have that huge modmore-sized values, do they? 
And - if you use the manager without this PR for a while, the data will be transfered anyway over the time, only in smaller, partly redundant chunks - which are producing an overall larger amount of traffic (because of the redundancy). Monthly quotas will surely not be affected to much (and - the lexicon data can be compressed quite effectively on the wire...).

What issues do you expect with loading all namespaces and topics? Any idea? I couldn't imagine any. The only one might be that it blows the modx object a bit. Can you measure that portion in your huge example instance?

christianseel commented:
From my perspective a larger file with better caching is much better than a smaller file that is requested multiple files… And I'd still say if you're using the current MODX manager on a mobile device, you will need a huge mobile data quota anyway…

jpdevries commented:
> From my perspective a larger file with better caching is much better than a smaller file that is requested multiple files… And I'd still say if you're using the current MODX manager on a mobile device, you will need a huge mobile data quota anyway…

https://github.com/modxcms/revolution/pull/12611 proves that a larger file with shared cache becomes more performant each time the user visits a new screen while optimizing for a single page experience bloats the payload as caches are not shared. I feel the research done there carries over here.

> I've tried this patch earlier on the modmore site (which has, for reasons including legacy and a lot of custom stuff, 48 namespaces) and the lang.js.php file went from 50kb (15kb gzipped) to 498kb (123kb gzipped) so it's approximately 10 times bigger in this particular example.

Ya but measure your ~50kb weight multiplied by the number of pages they visit. It could easily be larger and continue to grow as they click around. See https://github.com/modxcms/revolution/pull/12611#issuecomment-135659086 for a comparable analogy. 

The MODX Manager is not a website that prioritizes snappy initial loads – it is a web app. Optimization wise the two are polar opposites. We should be focusing on things like advising GZIP is enabled rather than reducing HTTP requests and optimizing for single page, first impression, visits.

OptimusCrime commented:
I too agree with one big file is better than multiple small ones. In terms of speed improvements, is there anything you can tell us? Any tests? Numbers?

Mark-H commented on connectors/lang.js.php:
```diff
@@ -5,68 +5,115 @@
+    $s = trim($s, ',');
+    $buf .= $s;
+
+    // use compressed js here - saves some bytes :-)
```
Kinda ironic that this PR increases the size of the lang.js ten-fold, but we need to use a compressed js snippet to save bytes?

rtripault replied:
Agreed. I'd rather be up for code readability instead!

christianseel committed:
create a proper json response and log errors

rtripault committed:
Make search bar work as expected on Chrome & Firefox

opengeek committed:
Remove all embedded escape characters when escape() is used on a string

This prevents SQL injection potential when a value is passed with an embedded escape character for the platform.

opengeek committed:
Merge remote-tracking branch 'xpdo/2.5.x' into 2.5.x

* xpdo/2.5.x:
  Remove all embedded escape characters when escape() is used on a string

opengeek committed:
Update xPDO from 2.5.x branch to get improved escape() behavior

Merge remote-tracking branch 'xpdo/2.5.x' into 2.5.x

* xpdo/2.5.x:
  Remove all embedded escape characters when escape() is used on a string

Jako committed:
Make search bar work as expected on Chrome & Firefox [#13405]

Merge remote-tracking branch 'upstream/pr/13405' into 2.5.x

* upstream/pr/13405:
  Make search bar work as expected on Chrome & Firefox

Mark-H committed:
[SECURITY-20] Prevent stored XSS in resource pagetitle

Reported by Tomáš Melicher via security@modx.com, ticket 20

opengeek committed:
Prevent stored XSS in resource pagetitle

Merge remote-tracking branch 'origin/pr/13415' into 2.5.x

* origin/pr/13415:
  [SECURITY-20] Prevent stored XSS in resource pagetitle

Mark-H committed:
[SECURITY-18] Fix stored XSS in user group name, and other potential manager XSS issues

Initial XSS report affecting the user group name and various places it gets rendered was from Anti Räis via security@modx.com, ticket 18 received April 3rd.

During the investigation of that report I found that ExtJS components that define custom tpls need the htmlEncode filter on possibly untrusted content. So along with fixing the reported issues, I've done a quick search for similar issues and also patched the potential issues I could find.

Mark-H committed:
Make sure the package provider attempts different methods of download packages if one fails #13417

Mark-H committed:
[SECURITY-20] Fix local file inclusion vulnerability in setup action parameter

Reported by Tomas Melicher via security@modx.com, ticket 20

Mark-H committed:
[SECURITY-19] Remove htaccess from allowed file types on new installations

Reported to security@modx.com by Anti Räis in ticket 19 and Tomáš Melicher in ticket 20, the ability to upload or create .htaccess files can cause code execution. Similar to how php files are not allowed out of the box, this patch prevents htaccess files by default to protect against that. Users that want to manage htaccess from the manager can still do so by editing the upload_files setting after installation.

Mark-H committed:
[SECURITY-20] Prevent reflected XSS in setup

Reported via security@modx.com by Tomáš Melicher. This patch makes sure configuration values, which may be provided by an attacker, are escaped before inserting them into the database configuration form of the setup.

Jako committed:
Remove htaccess from allowed file types on new installations

Merge remote-tracking branch 'upstream/pr/13423' into 2.5.x

* upstream/pr/13423:
  [SECURITY-19] Remove htaccess from allowed file types on new installations

Jako committed:
Fix local file inclusion vulnerability in setup action parameter

Merge remote-tracking branch 'upstream/pr/13422' into 2.5.x

* upstream/pr/13422:
  [SECURITY-20] Fix local file inclusion vulnerability in setup action parameter

Jako committed:
Prevent reflected XSS in setup

Merge remote-tracking branch 'upstream/pr/13424' into 2.5.x

* upstream/pr/13424:
  [SECURITY-20] Prevent reflected XSS in setup

Jako committed:
Proper use of json_encode and error handling for outputArray() in processors

Merge remote-tracking branch 'upstream/pr/13389' into 2.5.x

* upstream/pr/13389:
  create a proper json response and log errors

opengeek committed:
[SECURITY-20] Prevent XSS by cache poisoning via Host header

opengeek committed:
Prevent XSS cache poisoning via Host header

Merge remote-tracking branch 'origin/pr/13426' into 2.5.x

* origin/pr/13426:
  [SECURITY-20] Prevent XSS by cache poisoning via Host header

Mark-H committed:
Use modx_charset setting instead of hardcoding UTF-8

opengeek committed:
Improve local file inclusion protections

The existing protections would not work on Windows platforms with backslash path delimiters. This commit improves the LFI protections throughout the core to remove any sequence of 2 or more `.` characters regardless of the path delimiter that precedes or follows it.

opengeek committed:
Prevent user/email enumeration in forgot password feature

Addresses issue #13408

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/13428' into 2.5.x

* upstream/pr/13428:
  Improve local file inclusion protections

opengeek committed:
Prevent stored XSS in UserGroup names and various other fields [#13418]

Merge remote-tracking branch 'origin/pr/13418' into 2.5.x

* origin/pr/13418:
  Use modx_charset setting instead of hardcoding UTF-8
  [SECURITY-18] Fix stored XSS in user group name, and other potential manager XSS issues

opengeek committed:
Update minified js

opengeek committed:
Try all available methods when attempting to download transport packages [#13419]

Merge remote-tracking branch 'origin/pr/13419' into 2.5.x

* origin/pr/13419:
  Make sure the package provider attempts different methods of download packages if one fails #13417

opengeek committed:
Update lexicons from Crowdin

Crowdin project: http://translate.modx.com
Thanks to all translators and proofreaders for contribution!

opengeek committed:
Update version for 2.5.7-pl release

opengeek committed:
Remove unused path-search and url_search processors

opengeek committed:
Update version for 2.5.8-dev

Jako committed:
Fix logging an empty value

opengeek committed:
Add missing return statements to built-in validation classes

opengeek committed:
Merge remote-tracking branch 'xpdo/2.5.x' into 2.5.x

* xpdo/2.5.x:
  Add missing return statements to built-in validation classes

opengeek committed:
Update xPDO to fix issue with validation classes

Merge remote-tracking branch 'xpdo/2.5.x' into 2.5.x

* xpdo/2.5.x:
  Add missing return statements to built-in validation classes

theboxer committed:
Fix logging an empty value

* origin/pr/13445:
  Fix logging an empty value

theboxer committed:
Remove unused path_search and url_search processors

* origin/pr/13433:
  Remove unused path-search and url_search processors

rtripault committed:
Refresh the parent (resource) node when creating the first children

Finetuned committed:
Fixed issue #4581 #modxbughunt

theboxer committed:
Refresh element in tree after changing name in element's panel

* origin/pr/13502:
  Fixed issue #4581 #modxbughunt

Commit:
fixes issue #13505

theboxer committed:
Add view_template:true for the "Content Editor" access policy

* origin/pr/13508:
  fixes issue #13505

Commit:
Fix date format for created field of package in the package provider

Finetuned committed:
Fixed issue #11954 #modxbughunt

theboxer committed:
Add a mouseout listener to the 'Clear Filter' buttons across the manager

* origin/pr/13510:
  Fixed issue #11954 #modxbughunt

wuuti commented:
@Mark-H @rtripault Will this ever make it into the core? Thought it was quite useful and improves performance... I don't have any performance statistics, but it seems obvious that it is better in performance.
#modxbughunt

rtripault commented:
Not opposed to the idea here (but i'm not alone), according you "un-minify" the `_` function

wuuti committed:
Merge remote-tracking branch 'upstream/2.5.x' into improve-lexicon

wuuti committed:
Unminified _ function again.

wuuti committed:
Unminified _ function again.

wuuti commented:
Hm.... merging that old PR with 2.5.x seems not a good idea... any advice?

wuuti committed:
Merge remote-tracking branch 'origin/improve-lexicon' into improve-lexicon

opengeek commented:
Do not ever merge old PR's. You need to rebase your work on top of the branch you want to target.

OptimusCrime commented:
Is it fixed now @opengeek ? The PR contains commits from two users, but it looks correct. Could need a squash though.

sdrenth commented:
Tested PR and it looks good to go!

Mark-H commented:
@wuuti Would you mind re-creating this PR targeting 2.x instead? I tried changing the base but because 2.5.x and 2.x have diverged, it's not quite as straightforward for this to get merged into 2.x.

As a reminder:
2.x => new features/major improvements
2.MINOR.x (e.g. 2.5.x) => bug fixes for the 2.MINOR (e.g. 2.5) series

Commit:
Merge branch '2.x' into improve-lexicon

Conflicts:
	_build/build.sample.properties
	_build/build.xml
	core/docs/changelog.txt
	core/docs/version.inc.php
	core/model/modx/processors/security/group/getnodes.class.php
	core/model/modx/processors/security/user/get.class.php
	core/xpdo/changelog.txt
	manager/assets/modext/modx.jsgrps-min.js
	manager/assets/modext/workspace/package.browser.panels.js
	setup/includes/request/modinstallrequest.class.php

wuuti commented:
@Mark-H Not sure where the problem with the 2.x base right now is? Or do you want me to create a complete new PR? Hopefully this won't get forgotten again for 2 years... :-(

Mark-H commented:
@wuuti I see you tried to merge 2.5.x into 2.x, but that's unfortunately not the solution here as those branches diverged.. if you could create a new PR with just your improvements, that would be amazing.

wuuti commented:
Done in #13623
